logging.md

```diff
@@ -0,0 +1,196 @@
+
+## Pods / Containers Logs Locations
+
+Before diving into Gravity’s logging infrastructure, let’s explore how it is set up in Kubernetes in general.
+
+Kubernetes sets up symlinks in well-known locations for logs of all containers running in the cluster and groups them up in two directories on each node, by pod and by container. The directories where these logs go are `/var/log/pods` and `/var/log/containers` respectively.
+
+Note that these directories reside inside the planet container, not on the host.
+
+On each node, the logs of all containers running in the same pod will be grouped in that pod directory under `/var/log/pods`:
+
+```bash
+planet$ ls -l /var/log/pods/<pod-id>/
+```
+
+If we look at the logs of a particular container, we’ll see that these are in fact symlinks to the actual log files Docker keeps inside its data directory:
+
+```bash
+planet$ ls -l /var/log/pods/01146b3e-3709-11ea-b8d5-080027f6e425/init/
+total 4
+lrwxrwxrwx 1 root root 161 Jan 14 20:04 0.log -> /ext/docker/containers/236f...-json.log
+```
+
+In addition to `/var/log/pods`, Kubernetes also sets up a `/var/log/containers` directory which has a flat structure and the logs of all containers running on the node. The log files are also symlinks that point to the respective files in `/var/log/pods`:
+
+```bash
+planet$ ls -l /var/log/containers/
+total 180
+lrwxrwxrwx 1 root root 66 Jan 15 00:24 bandwagon-6c4b...-lqbll_kube-system_bandwagon-2641....log -> /var/log/pods/0b8e.../bandwagon/1.log
+```
+
+## Forwarder
+
+Log forwarder runs on every node of the cluster as a part of a DaemonSet:
+
+```bash
+$ kubectl -nkube-system get ds,pods -lname=log-forwarder
+```
+
+This component uses `remote_syslog2` to monitor files in the following directories:
+
+* `/var/log/containers/*.log`
+
+Like explained above, this directory contains logs for all containers running on the node.
+
+* `/var/lib/gravity/site/**/*.log`
+
+This directory contains Gravity-specific operation logs:
+
+```bash
+$ ls -l /var/lib/gravity/site/*/*.log
+```
+
+The forwarder Docker image and its configuration can be found here.
+
+## Collector
+
+Log collector is an rsyslogd server that’s running as a part of Deployment:
+
+```bash
+$ kubectl -nkube-system get deploy,pods -lrole=log-collector
+```
+
+The collector exposes rsyslog server via a Kubernetes Service where forwarders send entries from the files they monitor over tcp protocol:
+
+```bash
+$ kubectl -nkube-system get services/log-collector
+NAME            TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)                    AGE
+log-collector   ClusterIP   10.100.81.83   <none>        514/UDP,514/TCP,8083/TCP   6h54m
+```
+
+The collector itself writes all logs into `/var/log/messages` which is mounted into its pod so we can inspect all the logs on the node where the collector is running, inside the planet container:
```

Comment:
Worth noting a downside that if the collector `jumps`, the logs become scattered over multiple nodes.